File: examples/athletic-analyst-py/src/execution_trace.py
from __future__ import annotations

import json
import sys
import time
from datetime import datetime, UTC

from .test_config import TEST_MODE

EXECUTION_TRACE_PREFIX = "EXECUTION_TRACE: "


class ExecutionTracer:
    def __init__(self) -> None:
        self._enabled = TEST_MODE
        self._started_at = time.perf_counter()

    def trace(self, event: str, **details: object) -> None:
        if not self._enabled:
            return

        payload = {
            "event": event,
            "timestamp": datetime.now(UTC).isoformat().replace("+00:00", "Z"),
            "elapsed_ms": round((time.perf_counter() - self._started_at) * 1000),
            **details,
        }
        print(f"{EXECUTION_TRACE_PREFIX}{json.dumps(payload)}", file=sys.stderr, flush=True)

File: examples/athletic-analyst-ts/src/execution-trace.ts
import { TEST_MODE } from "./test-config.js";

export const EXECUTION_TRACE_PREFIX = "EXECUTION_TRACE: ";

export type ExecutionTraceEvent = {
  event: string;
  timestamp: string;
  elapsed_ms: number;
  [key: string]: unknown;
};

export type ExecutionTracer = {
  trace: (event: string, details?: Record<string, unknown>) => void;
};

export function createExecutionTracer(): ExecutionTracer {
  const start = Date.now();

  return {
    trace(event: string, details: Record<string, unknown> = {}): void {
      if (!TEST_MODE) return;

      const payload: ExecutionTraceEvent = {
        event,
        timestamp: new Date().toISOString(),
        elapsed_ms: Date.now() - start,
        ...details,
      };

      console.error(`${EXECUTION_TRACE_PREFIX}${JSON.stringify(payload)}`);
    },
  };
}
